### PR TITLE
Add an API to DDP for dynamically updating the underlying process group.

### PR DESCRIPTION
**Pull request by pritamdamania87**
# Motivation

If we would like to reinitialize DDP with a different PG with `torch.compile`, we need to do the following:

```
del old_ddp
del old_pg
pg = init_pg(...)
ddp = DDP(pg)
model = torch.compile(DDP)
```

This results in recompilation of the entire model and is very expensive. Since the only thing we need to update is the PG, we should be able to do this without having to compile the model again.

# Proposal

As a result, in this PR I've introduced an `_update_process_group` API which can dynamically update the underlying ProcessGroup used by DDP without needing to reinitialize DDP again.

cc @mrshenli @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

### CODE REVIEW

**Comment by XilunWu** [comment]
this PR has a lot of re-formatting which I suggest we exclude.

**Comment by pritamdamania87** [comment]
> this PR has a lot of re-formatting which I suggest we exclude.

@XilunWu This came from `lintrunner -a` automatically. Not sure if lint will fail if I don't apply those changes.

File: torch/_C/_distributed_c10d.pyi
Comment (wconstab) [inline torch/_C/_distributed_c10d.pyi:68]: i'm curious why you added force_bucket_rebuild.  Does this help solve the long-standing issue of compile not working with DDP(static_graph=True)?  If so (or even if its only part of the solution) it'd be good to know so we can get the rest of it solved and socialize that it's fixed

**Comment by wconstab** [comment]
> > this PR has a lot of re-formatting which I suggest we exclude.
> 
> @XilunWu This came from `lintrunner -a` automatically. Not sure if lint will fail if I don't apply those changes.

actually while i hate to be pedantic i think in this case since there are so many lines of code affected, it'd be worth splitting the PR into a stack where the first PR just applies lintrunner and the second PR just has your 'real' changes.

**Comment by wconstab** [comment]
@rohan-varma @fegin any issue with this PR?  (I can do a more detailed review if you dont want to but any objection on the addition of these APIs in principle?)

File: torch/_C/_distributed_c10d.pyi
Comment (pritamdamania87) [reply torch/_C/_distributed_c10d.pyi:68]: I did add a comment about this in the `_update_process_group` method of distributed.py. The reason is that rebuilding buckets incurs extra collectives. Now if the process group is updated such that some ranks have rebuilt buckets and other haven't. In that case there would be a hang since the ones that haven't yet rebuilt the buckets would do extra collectives.

Comment (pritamdamania87) [reply torch/_C/_distributed_c10d.pyi:68]: > Does this help solve the long-standing issue of compile not working with DDP(static_graph=True)? If so (or even if its only part of the solution) it'd be good to know so we can get the rest of it solved and socialize that it's fixed

I'm not sure about this and not sure if this might help with that issue since this wasn't the goal of this PR and I haven't tested that use case myself.

**Comment by pritamdamania87** [comment]
> actually while i hate to be pedantic i think in this case since there are so many lines of code affected, it'd be worth splitting the PR into a stack where the first PR just applies lintrunner and the second PR just has your 'real' changes.

So I had to run `lintrunner -a` since `make lint` was failing locally for me. If I remove the formatting changes won't lint fail on this PR and as a result block merge?

File: torch/csrc/distributed/c10d/init.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:612]: These are the only real changes in this file

**Comment by wconstab** [comment]
> > actually while i hate to be pedantic i think in this case since there are so many lines of code affected, it'd be worth splitting the PR into a stack where the first PR just applies lintrunner and the second PR just has your 'real' changes.
> 
> So I had to run `lintrunner -a` since `make lint` was failing locally for me. If I remove the formatting changes won't lint fail on this PR and as a result block merge?

my suggestion was to create a 'stack' where your first pr does the lint changes and then your second PR which is 'on top' does the changes.  (are you familiar with ghstack?  `pip install ghstack` and instructions on its github page)

**Comment by pritamdamania87** [comment]
> my suggestion was to create a 'stack' where your first pr does the lint changes and then your second PR which is 'on top' does the changes. (are you familiar with ghstack? pip install ghstack and instructions on its github page)

Looks like maybe there was something broken in my local setup causing the linter to change files. Removing the formatting still passes CI though

File: torch/csrc/distributed/c10d/reducer.cpp
Comment (wconstab) [inline torch/csrc/distributed/c10d/reducer.cpp:2314]: should we validate that the new PG matches the old PG in important ways (such as size)?

File: torch/nn/parallel/distributed.py
Comment (wconstab) [inline torch/nn/parallel/distributed.py:2259]: mby this is good enough for validation.  i'm actually wondering, is it invalid to change worldsize using the api or is that arguably valid?  (i guess we could use it to implement elastic tolerance for a dropped rank)

Comment (pritamdamania87) [reply torch/nn/parallel/distributed.py:2259]: > i'm actually wondering, is it invalid to change worldsize using the api or is that arguably valid? (i guess we could use it to implement elastic tolerance for a dropped rank)

This is actually the primary goal of the API, we want to change the worldsize. If you see the unit test below, we change the world size multiple times and ensure that DDP works appropriately and doesn't incur any compile overhead.

**Review by fduwjj — APPROVED** [review]
Overall, this looks good to me. Unblock for now but one thing I am not sure is that, are the changes in the reducer the only places we need to change? tbh, I am not reading reducer's code enough. Maybe, can @rohan-varma or @fegin also take a look? I also added more CI lables.

**Pull request closed without merging**
